### ISSUE TITLE
voicefocus is not working in mobile chrome > 96

### ISSUE THREAD

**Issue by husaindevelop** [aws/amazon-chime-sdk-js#2291]
### What happened and what did you expect to happen?

We are using amazon chime single js sdk latest version to create meetings. Under it, the feature amazon voice focus was configured and is successfully working in laptop device, but unfortunately mobile (smartphone) chrome does not initialize the feature as expected, though it logs that voicefocus is supported but the voice focus does not start unlike the laptop.

### Have you reviewed our existing documentation?

- [X] GitHub [issues](https://github.com/aws/amazon-chime-sdk-js/issues)
- [X] FAQs [page](https://aws.github.io/amazon-chime-sdk-js/modules/faqs.html)
- [X] README [page](https://github.com/aws/amazon-chime-sdk-js#amazon-chime-sdk-for-javascript)
- [X] Feature-specific [guides](https://github.com/aws/amazon-chime-sdk-js#guides)

### Reproduction steps

Please be informed, we are using amazon chime single js sdk latest version. With help of your team, we were able to successfully implement amazon voice focus on laptop. It successfully gives us success message on initialization of voice focus.
```
[2022-06-11T06:18:02.984Z [INFO] MyLogger - Decided execution approach {"supported":true,"useSIMD":true,"processor":"voicefocus-inline-processor","executionApproach":"inline","variant":"c50","executionQuanta":3}
vp00017.php?v=Yes:86 Amazon Voice Focus enabled]
```

But same with mobile chrome device, amazon voice focus is supported with affirmation in logs, but subsequently unlike the laptop, the script here dies with the following line
 `const vfDevice = await transformer.createTransformDevice(chosenAudioInput);`

Section of code which includes amaonz voice focus
```
 (async () => {
            const Logger = new ChimeSDK.ConsoleLogger('MyLogger', ChimeSDK.LogLevel.INFO);
    
           
try {
    //const transformer = await ChimeSDK.VoiceFocusDeviceTransformer.create({}, { logger });
    const isVoiceFocusSupported = await ChimeSDK.VoiceFocusDeviceTransformer.isSupported();
    const vs = await ChimeSDK.supportsSetSinkId;
            
    
} catch (e) {
    console.error(e);
    
  // Will only occur due to invalid input or transient errors (e.g., network).
  isVoiceFocusSupported = false;
  vs = false;
}
           
            const isVoiceFocusSupported = await ChimeSDK.VoiceFocusDeviceTransformer.isSupported();
            this.deviceController = new ChimeSDK.DefaultDeviceController(Logger, { enableWebAudio: isVoiceFocusSupported });
            const transformer = await ChimeSDK.VoiceFocusDeviceTransformer.create({}, { logger: Logger });
            const chosenAudioInput = audioInputDevices[0].deviceId;    
            const vfDevice = await transformer.createTransformDevice(chosenAudioInput);
 const deviceToUse = vfDevice || chosenAudioInput;
await deviceController.chooseAudioInputDevice(deviceToUse);

if (vfDevice) {
  console.log('Amazon Voice Focus enabled');
}
else {
    console.log('Amazon Voice Focus disabled');
}
```

Logs are attached herealong

### Amazon Chime SDK for JavaScript version

latest

### What browsers are you seeing the problem on?

chrome

### Browser version

chrome 97

### Browser console logs

[logs.txt](https://github.com/aws/amazon-chime-sdk-js/files/8883385/logs.txt)

**Comment by LeviSklaroff** [comment]
- You mentioned that it logs voice focus is supported, but I don't see anything in the given logs indicating this. The isSupported function will not always log something or throw an error when it returns false.
- Could you provide more details about the device this is on?

**Comment by husaindevelop** [comment]
Ok, so i have edited the code a little to emit whether voice focus is supported vide the following code:

```
try {
    const isVoiceFocusSupported = await ChimeSDK.VoiceFocusDeviceTransformer.isSupported();
    const vs = await ChimeSDK.supportsSetSinkId;
    console.log ("Voice focus "+isVoiceFocusSupported);

} catch (e) {
    console.error(e);
    
  // Will only occur due to invalid input or transient errors (e.g., network).
  isVoiceFocusSupported = false;
  vs = false;
  console.log ("Voice focus"+ isVoiceFocusSupported);

}
           
```
It does return true. Other than that, all other parameters do not emit any error except for this line
`const vfDevice = await transformer.createTransformDevice(chosenAudioInput,{ logger: Logger });`
This code returns "undefined" response in the mobile device, 
whereas it returns the following response when we try it with laptop:
"MyLogger - Choosing transform input device [object Object]" 

We fail to understand as why the aforesaid code works in laptop but not on a mobile device.

**Comment by LeviSklaroff** [comment]
@husaindevelop I've deleted your comment because you posted a public meeting url, we ask that you do not post these on here as anyone can access and start a meeting using the url. I will investigate into this though, from what I can tell your code looks correct. The isSupported for voice focus does have some limitations and its possible voice focus might not work for your device, could you try testing it on our browser demo? I'll also try seeing if I might be able to test a Redmi device.

**Comment by husaindevelop** [comment]
Sorry i didn't see your comment. Its working in browser demo. Not on redmi 9A

**Comment by husaindevelop** [comment]
Any update on this issue ?

**Comment by richhx** [comment]
Hi @husaindevelop ,

Looking at the `logs.txt`, I see the following:
```
ConsoleLogger.js:76 2022-06-11T06:10:05.444Z [INFO] MyLogger - Choosing intrinsic audio input device default
ConsoleLogger.js:76 2022-06-11T06:10:05.446Z [INFO] MyLogger - requesting new audio device with constraint {"audio":{"echoCancellation":true,"googEchoCancellation":true,"googEchoCancellation2":true,"googAutoGainControl":true,"googAutoGainControl2":true,"googNoiseSuppression":true,"googNoiseSuppression2":true,"googHighpassFilter":true,"deviceId":{"exact":"default"},"sampleRate":{"ideal":48000},"sampleSize":{"ideal":16},"channelCount":{"ideal":1}}}
ConsoleLogger.js:76 2022-06-11T06:10:05.914Z [INFO] MyLogger - got audio device for constraints {"audio":{"echoCancellation":true,"googEchoCancellation":true,"googEchoCancellation2":true,"googAutoGainControl":true,"googAutoGainControl2":true,"googNoiseSuppression":true,"googNoiseSuppression2":true,"googHighpassFilter":true,"deviceId":{"exact":"default"},"sampleRate":{"ideal":48000},"sampleSize":{"ideal":16},"channelCount":{"ideal":1}}}
ConsoleLogger.js:76 2022-06-11T06:10:05.925Z [INFO] MyLogger - API/DefaultDeviceController/chooseAudioInputDevice "default" -> "success"
ConsoleLogger.js:76 2022-06-11T06:10:05.926Z [INFO] MyLogger - API/DefaultAudioVideoFacade/3d5e2c86-959e-42c3-b6d4-e33297510706/0fd36794-5517-d556-7c43-3301c1bcced8/chooseVideoInputQuality {"width":"400","height":"700","frameRate":"15 fps","maxBandwidthKbps":"1440 kbps"}
ConsoleLogger.js:76 2022-06-11T06:10:05.928Z [INFO] MyLogger - Choosing intrinsic audio input device default
ConsoleLogger.js:76 2022-06-11T06:10:05.931Z [INFO] MyLogger - reusing existing audio input device
ConsoleLogger.js:76 2022-06-11T06:10:05.932Z [INFO] MyLogger - API/DefaultDeviceController/chooseAudioInputDevice "default" -> "success"
vp00017.php?v=Yes:89 Amazon Voice Focus disabled
```

The key line here is "Choosing intrinsic audio input device default" indicating that the device being passed is the default device (`chosenAudioInput`) and not the Voice Focus device (`vfDevice`).  As you mention, `VoiceFocusDeviceTransformer` is not initializing the Voice Focus device, which could be for a variety of reasons.

As @LeviSklaroff mentioned, I don't actually see indication from `logs.txt` saying Voice Focus is actually supported on the mobile device. It runs the estimates, but running the estimates does not necessarily mean that it is supported. If you log the value returned from `const isVoiceFocusSupported = await ChimeSDK.VoiceFocusDeviceTransformer.isSupported();`, my assumption is that it is `false`. 

If you want to see more information about why it's `false`, you can pass the `logger` as an options parameter to `VoiceFocusDeviceTransformer.isSupported(undefined, { logger})`. In addition, you can change the logging level to `LogLevel.DEBUG`. If you are seeing the supported value as `true`, please attach the logs at debug level and I can give it a further look.

**Comment by husaindevelop** [comment]
Shall check and revert.

**Comment by husaindevelop** [comment]
Did just as you mentioned, changed ChimeSDK.LogLevel to DEBUG level, updated the following lines of code for better debugging.

```
const isVoiceFocusSupported = ChimeSDK.VoiceFocusDeviceTransformer.isSupported(undefined, {logger: Logger});
const vs = await ChimeSDK.supportsSetSinkId;
console.log("VF supported "+isVoiceFocusSupported);
```
We got the following results in the log. Its attached herealong for your information
[log2.txt](https://github.com/aws/amazon-chime-sdk-js/files/9007975/log2.txt)

Based on immediate look, this line does return true. 
`console.log("VF supported "+isVoiceFocusSupported);`

Please take a look at the logs and see how we can resolve this. Thanking you in advance.

**Comment by richhx** [comment]
Hmm looking at the logs2.txt, and I don't see it returning true. I see the following line:

```
video_chat3.php?v=Yes:60 VF supported [object Promise]
```

Can you provide more information as to where it's logging true? 

Note that even if Voice Focus says it's supported in that line, it can still fail to create the device due to other runtime constraints that it detects. This is noted in [code docs](https://github.com/aws/amazon-chime-sdk-js/blob/main/src/voicefocus/VoiceFocusDeviceTransformer.ts#L73-L75) at the moment, but we can look into making it easier to find. 

I should've told you this earlier, but if you check `transformer.isSupported()`, that is actually the [better indicator](https://github.com/aws/amazon-chime-sdk-js/blob/main/src/voicefocus/VoiceFocusDeviceTransformer.ts#L125-L128) as to whether or not it's supported whereas the static `isSupported()` call is meant as a quick check to prune the majority of devices.

Looking at the `log2.txt`, I see that the SIMD timing is 46.86 which is pretty close to the border of not being supported based on the default execution preferences. Even if it does have isSupported set to true or is forced to run, it is likely it will not work well due to the performance limitations.

TL;DR: I still believe it's not supported on the mobile device you're testing on due to performance limitations.

**Comment by husaindevelop** [comment]
Voicefocus support does return true vide this line `video_chat3.php?v=Yes:62 VF supported true`
The code we used to check if voice focus is supported or not:
  `const isVoiceFocusSupported1 = await ChimeSDK.VoiceFocusDeviceTransformer.isSupported();`

However you are right. the transformer.isSupported does return false, which effectively means voicefocus is not supported on my device.

We have few questions.
1. Now by "Performance limitations" what do you actually mean. 
2. What is the configuration a mobile device should have to actually support voice focus, 
3. Alternatively is it because there is high usage of CPU in the device that disables voice focus. 
4. And without voice focus and a video quality of `meetingSession.audioVideo.chooseVideoInputQuality(1280,720,30,2500);`, it throws up pretty big squeaking noise, almost to the intolerable levels. In this scenario, video call is big no-no

Guide us as how we got about on this issue

**Comment by richhx** [comment]
1. Amazon Voice Focus requires that the device is able to process a render quantum of audio in a given amount of time. If the device is not capable of doing so, it will fall behind and not be able to execute leading to partially transformed audio. This partially transformed audio is not "clean" audio. As a result, we have an estimator that runs on a device to see whether or not Voice Focus can actually run to avoid the partially transformed audio. If it fails to meet the benchmarks, the transformer will refuse to create the Voice Focus transformer device.
2. It needs to be powerful enough to pass the aforementioned benchmarks in our code.
3. See response 1.
4/5. As you mentioned, this happens even without Voice Focus. The SDK team can assist you on the reasoning as to why it's not possible, but my guess is maybe also some performance limitation of the device not being able to keep up. Perhaps @LeviSklaroff can help out with more information.

**Comment by husaindevelop** [comment]
So how do we connect him. Or if you don't mind, could just put in a note to him to help us out with some information on this same ticket.

**Comment by husaindevelop** [comment]
Any update

**Comment by richhx** [comment]
When you say connect "him", are you referring to getting Voice Focus to work on mobile? As mentioned before, it will not work if the performance requirements are not met and we cannot really do anything about that at the moment.

**Comment by husaindevelop** [comment]
When i said connect him, i meant connecting leviSklaroff for more information. I made this point on the basis of your observation: "but my guess is maybe also some performance limitation of the device not being able to keep up. Perhaps @LeviSklaroff can help out with more information.".

**Comment by richhx** [comment]
@ltrung are you able to help out?

**Comment by ltrung** [comment]
@husaindevelop For 4 & 5, is it only for mobile or also web browser? Could you give more details on the type of devices you have the issues on?

**Comment by husaindevelop** [comment]
- No only for mobile browsers. 
- No, its just internal speakers
- Regarding type of devices, we have Redmi 9A, MIUI 12 and Android 10 version. We also tried on safari 12 on IPhone 6. It did not work there either. 

@ltrung, its ok if it does not work on our device Redmi 9A. Possibly as @richx mentioned, due to its device's incapability, it might not be working. But least we expect to know is what is the minimum configuration a device should carry to run the voicefocus, so that we can communicate the same to our users. Hopefully you might be able to give us a definitive answer.

**Comment by husaindevelop** [comment]
Any update ?

**Comment by xuesichao** [comment]
Hi @husaindevelop, we don't have a specific minimum device requirements. You can check our [Amazon Voice Focus Guide](https://aws.github.io/amazon-chime-sdk-js/modules/amazonvoice_focus.html#can-i-use-amazon-voice-focus-and-echo-reduction-in-my-application) as a reference.

>As such, not all mobile devices or lower-spec laptop or desktop computers will be sufficiently powerful, or will be able to use Amazon Voice Focus and Echo Reduction while also supporting multiple video streams and rich application functionality.
>
>The default configuration will adapt to available processor power and adjust quality accordingly, but some browsers and devices will simply be unable to enable the feature. Android browsers are theoretically compatible, but typically cannot meet the performance requirements. iOS 14 ships Safari 14, and relatively modern iPhones (e.g., iPhone X) have been found to be fast enough.

**Comment by husaindevelop** [comment]
Ok. Thanks, i will consider this ticket to be closed
